source/slang/slang-parser.cpp

```diff
@@ -5579,6 +5589,119 @@ namespace Slang
         return varDeclrStatement;
     }
 
+    static Expr* constructIfLetPredicate(Parser* parser, VarExpr* varExpr)
+    {
+        // create a "var.hasValue" expression
+        MemberExpr* memberExpr = parser->astBuilder->create<MemberExpr>();
+        memberExpr->baseExpression = varExpr;
+        parser->FillPosition(memberExpr);
+        memberExpr->name = getName(parser, "hasValue");
+
+        // create a "==" operator
+        auto opExpr = parser->astBuilder->create<VarExpr>();
+        opExpr->name = getName(parser, "==");
+        opExpr->scope = parser->currentScope;
+        parser->FillPosition(opExpr);
+
+        // create a 'true' literal
+        BoolLiteralExpr* boolLiteralExpr = parser->astBuilder->create<BoolLiteralExpr>();
+        boolLiteralExpr->value = true;
+
+        // create a "var.hasValue == true" expression
+        InfixExpr* infixExpr = parser->astBuilder->create<InfixExpr>();
+        infixExpr->loc = opExpr->loc;
+        infixExpr->functionExpr = opExpr;
+        infixExpr->arguments.add(memberExpr);
+        infixExpr->arguments.add(boolLiteralExpr);
+        return infixExpr;
+    }
+
+    // Parse the syntax 'if (let var = X as Y)'
+    Stmt* Parser::parseIfLetStatement()
+    {
+        ScopeDecl* scopeDecl = astBuilder->create<ScopeDecl>();
+        pushScopeAndSetParent(scopeDecl);
+
+        SeqStmt* newBody = astBuilder->create<SeqStmt>();
+
+        IfStmt* ifStatement = astBuilder->create<IfStmt>();
+        FillPosition(ifStatement);
+        ReadToken("if");
+        ReadToken(TokenType::LParent);
+
+        // parse 'let var = X as Y'
+        ReadToken("let");
+        auto identifierToken = ReadToken(TokenType::Identifier);
+        ReadToken(TokenType::OpAssign);
+        auto initExpr = ParseInitExpr();
+
+        // insert 'let tempVarDecl = X as Y;'
+        auto tempVarDecl = astBuilder->create<LetDecl>();
+        tempVarDecl->nameAndLoc = NameLoc(getName(this, "OptVar"), identifierToken.loc);
+        tempVarDecl->initExpr = initExpr;
+        AddMember(currentScope->containerDecl, tempVarDecl);
+
+        DeclStmt* tmpVarDeclStmt = astBuilder->create<DeclStmt>();
+        FillPosition(tmpVarDeclStmt);
+        tmpVarDeclStmt->decl = tempVarDecl;
+        newBody->stmts.add(tmpVarDeclStmt);
+
+        // construct 'if (tempVarDecl.hasValue == true)'
+        VarExpr* tempVarExpr = astBuilder->create<VarExpr>();
+        tempVarExpr->scope = currentScope;
+        FillPosition(tempVarExpr);
+        tempVarExpr->name = tempVarDecl->getName();
+        ifStatement->predicate = constructIfLetPredicate(this, tempVarExpr);
```

Comment:
We are not strictly mimicking C++ syntax, but in C++, this is possible,
```
#include <cstdio>
int main()
{
        if (int a = 0; a > 0)
        {
                printf("a\n");
        }
        else if (a == 0) // <== available on the else clause.
        {
                printf("b\n");
        }
        return 0;
}
```
It seems like your implementation would work for this scoping.
Can we have some test cases for it?

Reply:
this implementation actually doesn't work in else scope. The variable only works in `if` branch.
line 5692 guarantees that.

But this is still a design decision, the variable will be `none` in else branch.

Reply:
The tests cases will be uploaded.

Reply:
The if following else will be another statement handled by the parse statement so that should be fine.